### PR TITLE
Major upgrade merge main

### PR DESCRIPTION
**Pull request by halsaphi**
Merging main into major-upgrade to sync major-upgrade branch with latest changes on main. In preparation to releasing the updated docs for the hard fork.

Please note: Have removed info boxes stating zkApps are not available on mainnet.

Have fixed broken links and accepted changes from main to fix conflicts.

### CODE REVIEW

File: docs/exchange-operators/faq.mdx [hunk 1]
```diff
@@ -54,7 +54,7 @@ The max mempool size is 3,000. After it hits that size, transactions with the lo
 
 Set your fee to an amount higher than 0.001 MINA, the current average fee for transactions in the pool.
 
-You can view the fees for pending transactions and adjust your fees accordingly: [https://minaexplorer.com/pending-transactions](https://minaexplorer.com/pending-transactions).
+You can view the fees for pending transactions and adjust your fees accordingly: [https://minascan.io/devnet/txs](https://minascan.io/devnet/txs).
```
Comment (kaozenn) [inline docs/exchange-operators/faq.mdx:57]: Should it be mainnet instead of devnet?

Comment (halsaphi) [reply docs/exchange-operators/faq.mdx:57]: should probably be - https://minascan.io/mainnet/txs/pending-txs - I'll update

Comment (halsaphi) [reply docs/exchange-operators/faq.mdx]: I've updated + made the rosetta section callapsible so it matches the other doc sections.

File: docs/exchange-operators/faq.mdx [hunk 2]
```diff
@@ -152,23 +152,11 @@ Use a recursive query. See [Query the database](/node-operators/archive-node/get
...
-```
-5f704cc0c82e0ed70e873f0893d7e06f148524e3f0bdae2afb02e7819a0c24d1
-```
-
-Devnet:
-
-```
-8af43cf261ea10c761ec540f92aafb76aec56d8d74f77c836f3ab1de5ce4eac5
-```
+To check whether you are running a Devenet or Mainnet build, run `Mina client status`. Next, compare the output's chain ID of your node to the expected chain ID of the network you are trying to connect to. You can find required information for comparison within the [GitHub announcements](https://github.com/MinaProtocol/mina/discussions/categories/announcements) or [Discord](https://discord.com/channels/484437221055922177/601171209287368715) server.
```
Comment (Trivo25) [inline docs/exchange-operators/faq.mdx:155]: `Devenet`

File: docs/exchange-operators/rosetta/samples/requests.mdx [hunk 3]
```diff
@@ -34,13 +56,49 @@ def _request(path, data=None):
...
+
+```ts
+function makeBlockIdentifier(idOrHash: number | string) {
+    const identifierKey = (typeof idOrHash === "number") ? "index" : "hash"
+    return {
+        block_identifier: {[identifierKey]: idOrHash}
+    }
+}
+
+function makeTxIdentifier(hash: string) {
```
Comment (Trivo25) [inline docs/exchange-operators/rosetta/samples/requests.mdx:80]: I suggest cleaning this cope snippet up a little bit

File: docs/exchange-operators/rosetta/samples/requests.mdx [hunk 4]
```diff
@@ -197,6 +258,56 @@ According to the Rosetta specification, the array of `operation` objects must be
 
 To implement a helper function to construct the operations array for a MINA token transfer:
 
+<Tabs>
+<TabItem value="ts" label="TypeScript">
+
+```ts
+function makeTransferPayload(from: string, to: string, feeNano: number, valueNano: number) {
```
Comment (Trivo25) [inline docs/exchange-operators/rosetta/samples/requests.mdx:265]: same here

File: docs/exchange-operators/rosetta/samples/requests.mdx [hunk 5]
```diff
@@ -241,10 +352,109 @@ def _make_transfer_payload(address_from, address_to, fee_nano, value_nano):
...
+}
+
+async function txPayloads(srcPublicKey: string, srcAddress: string, destAddress: string, feeNano: number, valueNano: number) {
+    // If fee_nano is undefined, it will get suggested fee from /construction/metadata response
+    const meta = await txMetadata(
+        srcPublicKey, srcAddress, destAddress, feeNano, valueNano)
+
+    if (feeNano === 0)
+        feeNano = meta.suggested_fee[0].value
+    console.log(feeNano)
```
Comment (Trivo25) [inline docs/exchange-operators/rosetta/samples/requests.mdx:427]: is this log supposed to stay here?

File: docs/exchange-operators/rosetta/samples/scan-blocks.mdx [hunk 6]
```diff
@@ -21,8 +43,29 @@ def wait_for_block(block_index):
...
 
+<Tabs>
+<TabItem value="ts" label="TypeScript">
+
+```ts
+let latestBlockHeight = (await network_status()).current_block_identifier.index
+while (true) {
+    const lastBlock = waitForBlock(latestBlockHeight)
+    
+    # some processing according to business logic
```
Comment (Trivo25) [inline docs/exchange-operators/rosetta/samples/scan-blocks.mdx:59]: this is python comment in ts code

File: docs/node-developers/graphql-api.mdx [hunk 7]
```diff
@@ -138,9 +138,10 @@ subscription {
 
 The new block subscription can also be limited to return only new blocks created by a defined public key with the `publicKey` argument.
 
-### GraphQL Public Proxy
+### GraphQL API and public Internet
 
-While exposing the GraphQL API to the internet is not recommended, an implementation of a [GraphQL proxy](https://github.com/MinaProtocol/mina/tree/master/automation/services/graphql-public-proxy) is available that removes the queries that pose a threat, but still allows querying of the node's data from a public endpoint.
+Exposing the GraphQL endpoint with full API support to the public Internet **is not** recommended.
```
Comment (Trivo25) [inline docs/node-developers/graphql-api.mdx:143]: Would be worth noting _why_ exactly it is not recommended, disclose possible vulnerabilities

File: docs/zkapps/faq.mdx [hunk 8]
```diff
@@ -141,7 +142,11 @@ Use the `requireSignature` command. See the [requireSignature](/zkapps/o1js-refe
 
 ### How do I configure who has the authority to interact and make changes to a specific part of an account?
 
+<<<<<<< HEAD
```
Comment (Trivo25) [inline docs/zkapps/faq.mdx:145]: this looks like a unresolved merge conflict

File: docs/zkapps/o1js-reference/README.mdx [hunk 9]
```diff
@@ -0,0 +1,219 @@
+---
```
Comment (Trivo25) [inline docs/zkapps/o1js-reference/README.mdx:1]: I suggest updating the o1js reference with the latest version,

File: docs/zkapps/o1js/index.mdx [hunk 10]
```diff
@@ -36,18 +36,31 @@ function knowsPreimage(preimage: Field) {
...
+
+## Audits of o1js
+
+* **Third-party audit (Q1/Q2 2024)**. An audit by an external security firm is ongoing and on track to complete in Q2 2024. Results will be shared when available.
+
+* [**Internal audit (Q1 2024)**](https://github.com/o1-labs/o1js/files/15192821/Internal.o1js.audit.Q1.2024.pdf). In March 2024, the o1js team spent roughly two person-weeks to conduct an internal audit of parts of the o1js code base. The audit focused on reviewing core provable code. A number of issues were found and fixed.
+
+Please see our page on [Security and zkApps](/zkapps/writing-a-zkapp/introduction-to-zkapps/secure-zkapps) for more information on ensuring your zkApp is secure.
 
+<<<<<<< HEAD:docs/zkapps/o1js/index.md
```
Comment (Trivo25) [inline docs/zkapps/o1js/index.mdx:63]: unresolved merge conflict

**Review by Trivo25 — CHANGES_REQUESTED** [review]
I would suggest merging o1labs/main into one of the branches to get the most up to date changes, not sure when this was last done but we are adding to the main branch quite frequently at the moment (including quite significant structural changes). Additionally, there are a few unresolved "merge conflicts" and we should agree on a style guide for the Rosetta code snippets (eg `camelCase` or `snake_case`

File: docs/zkapps/faq.mdx (hunk 8 above)
Comment (halsaphi) [reply docs/zkapps/faq.mdx]: fixed the missed merge conflicts - found 5

File: docs/zkapps/o1js/index.mdx (hunk 10 above)
Comment (halsaphi) [reply docs/zkapps/o1js/index.mdx]: fixed the missed merge conflicts - found 5

**Comment by qwadratic** [comment]
@Trivo25 correct me if I am wrong, but camelCase used in typescript samples and snake_case - in python. according to styles preferred by these communities

**Comment by Trivo25** [comment]
> @Trivo25 correct me if I am wrong, but camelCase used in typescript samples and snake_case - in python. according to styles preferred by these communities


Here are the two styles mixed up - `relatedIdxs` and `op_type` for example.
```ts
function makeOperation(idx: number, relatedIdxs: number[], op_type: string, addr: string, value: number, isPositive: boolean) {
```

Other than thatm I think it would be beneficial to let a linter run over the code examples, snippets like these

```ts
return {
        block_identifier: {[identifierKey]: idOrHash}
}
```

can be improved for better readability 

```ts
return {
    block_identifier: {
        [identifierKey]: idOrHash
    }
}
```

Sometimes spaces and indentations are off as well

**Comment by qwadratic** [comment]
ok I will do it tomo

File: docs/zkapps/o1js/index.mdx [hunk 11]
```diff
@@ -60,7 +60,4 @@ Start your o1js journey by learning about [basic zk programming concepts](/zkapp
 
 Please see our page on [Security and zkApps](/zkapps/writing-a-zkapp/introduction-to-zkapps/secure-zkapps) for more information on ensuring your zkApp is secure.
 
-<<<<<<< HEAD:docs/zkapps/o1js/index.md
 Start your o1js journey by learning about [basic zk programming concepts](./o1js/basic-concepts).
-=======
->>>>>>> main:docs/zkapps/o1js/index.mdx
```
Comment (mitschabaude) [inline docs/zkapps/o1js/index.mdx]: this conflict was fixed incorrectly, it reintroduces a sentence that is out of context in this location

Comment (halsaphi) [reply docs/zkapps/o1js/index.mdx]: fixed

File: docs/exchange-operators/rosetta/samples/requests.mdx (hunk 3 above)
Comment (halsaphi) [reply docs/exchange-operators/rosetta/samples/requests.mdx:80]: I'd suggest addressing these comments as a separate PR once main has been merged

File: docs/exchange-operators/rosetta/samples/requests.mdx (hunk 4 above)
Comment (halsaphi) [reply docs/exchange-operators/rosetta/samples/requests.mdx:265]: I'd suggest addressing these comments as a separate PR once main has been merged

File: docs/exchange-operators/rosetta/samples/requests.mdx (hunk 5 above)
Comment (halsaphi) [reply docs/exchange-operators/rosetta/samples/requests.mdx:427]: I'd suggest addressing these comments as a separate PR once main has been merged

File: docs/exchange-operators/rosetta/samples/scan-blocks.mdx (hunk 6 above)
Comment (halsaphi) [reply docs/exchange-operators/rosetta/samples/scan-blocks.mdx:59]: I'd suggest addressing these comments as a separate PR once main has been merged

**Comment by qwadratic** [comment]
@Trivo25 

Thank you for the suggestion, nice catch. While it is on my backlog, I suggest to merge this PR and do the change you suggested as a separate one. wdyt?

**Review by Trivo25 — APPROVED** [review]

**Pull request merged**